pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingAggregationDataTableReducer.java

```diff
@@ -0,0 +1,172 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.query.reduce;
+
+import com.google.common.base.Preconditions;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import org.apache.pinot.common.response.broker.AggregationResult;
+import org.apache.pinot.common.response.broker.BrokerResponseNative;
+import org.apache.pinot.common.response.broker.ResultTable;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.common.utils.DataSchema.ColumnDataType;
+import org.apache.pinot.common.utils.DataTable;
+import org.apache.pinot.core.query.aggregation.function.AggregationFunction;
+import org.apache.pinot.core.query.aggregation.function.AggregationFunctionUtils;
+import org.apache.pinot.core.query.request.context.QueryContext;
+import org.apache.pinot.core.transport.ServerRoutingInstance;
+import org.apache.pinot.core.util.QueryOptionsUtils;
+
+
+/**
+ * Helper class to reduce and set Aggregation results into the BrokerResponseNative
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class StreamingAggregationDataTableReducer implements StreamingReducer {
+  private final QueryContext _queryContext;
+  private final AggregationFunction[] _aggregationFunctions;
+  private final boolean _preserveType;
+  private final int _numAggregationFunctions;
+  private final Object[] _intermediateResults;
+
+  private DataSchema _dataSchema;
+  private DataTableReducerContext _dataTableReducerContext;
+
+  StreamingAggregationDataTableReducer(QueryContext queryContext) {
+    _queryContext = queryContext;
+    _aggregationFunctions = queryContext.getAggregationFunctions();
+    Map<String, String> queryOptions = queryContext.getQueryOptions();
+    _preserveType = QueryOptionsUtils.isPreserveType(queryOptions);
+
+    Preconditions.checkNotNull(_aggregationFunctions);
+    _numAggregationFunctions = _aggregationFunctions.length;
+    _intermediateResults = new Object[_numAggregationFunctions];
+    _dataSchema = null;
+  }
+
+  @Override
+  public void init(DataTableReducerContext dataTableReducerContext) {
+    _dataTableReducerContext = dataTableReducerContext;
+  }
+
+  /**
+   * Reduces data tables and sets aggregations results into
+   * 1. ResultTable if _responseFormatSql is true
+   * 2. AggregationResults by default
+   */
+  @Override
+  public void reduce(ServerRoutingInstance key, DataTable dataTable) {
```

Comment:
(major) We need to make sure all `reduce()` is thread safe. Same for other streaming reducers.

Reply:
lol yeah this PR is not ready for review yet. just opening this draft to run CI :-P